File: keggAPI.R
#if(!requireNamespace('BiocManager', quietly = TRUE))      
#  install.packages('BiocManager') 
#BiocManager::install('KEGGREST') 

library('KEGGREST')

pathways <- keggList('pathway')

res <- data.frame('ko' = 0, 'name' = 0, 'description' = 0, 'module_id' = 0, 'module' = 0, 'pathway_id' = 0, 'pathway' = 0, 'class' = 0)
res <- res[-1, ]

# 遍历所有map
for(i in 1: length(strsplit(pathways, '\n'))){
  pathway <- strsplit(pathways, '\n')[i]
#  print(pathway)}
  map_entry <- names(pathway)
#  print(map_entry)}
  map_name <- pathway[[1]]
  map_info <- keggGet(map_entry)
  map_id <- map_info[[1]]$KO_PATHWAY
  # 处理缺少分类的情况
  if(length(map_info[[1]]$CLASS != 0)){
    map_class <- map_info[[1]]$CLASS
  }
  else{
    map_class <- 'None'
  }
  # 提取map中的模块
  modules <- map_info[[1]]$MODULE
  # 处理不包含模块的map
  if(length(modules) != 0){
    # 遍历map中的模块
    for(j in 1: length(strsplit(modules, '\n'))){
      module <- strsplit(modules, '\n')[j]
      module_id <- names(module)
      module_name <- module[[1]]
      module_entry <- paste('module:', module_id, sep = '')
      module_info <- keggGet(module_entry)
      orthologies <- module_info[[1]]$ORTHOLOGY
      # 遍历模块中的ko
      for(k in 1: length(strsplit(orthologies, '\n'))){
        ko <- strsplit(orthologies, '\n')[k]
        # 处理多个ko对应同一功能的情况
        if(length(strsplit(names(ko), ',')[[1]]) > 1){
          for(kos in strsplit(names(ko), ',')){
            # 处理ko编号带有其它信息的情况
            if(startsWith(kos, 'K')){
              ko_entry <- paste('orthology:', substr(kos, 1, 6), sep = '')
              ko_info <- keggGet(ko_entry)
              ko_name <- ko_info[[1]]$NAME
              ko_description <- ko_info[[1]]$DEFINITION
              temp <- data.frame('ko' = kos, 'name' = ko_name, 'description' = ko_description, 'module_id' = module_id, 'module' = module_name, 'pathway_id' = map_id, 'pathway' = map_name, 'class' = map_class)
              res <- rbind(res, temp)
            }
            else{
              next
            }
          }
        }else{
          ko_id <- names(ko)
          ko_entry <- paste('orthology:', ko_id, sep = '')
          ko_info <- keggGet(ko_entry)
          ko_name <- ko_info[[1]]$NAME
          ko_description <- ko_info[[1]]$DEFINITION
          temp <- data.frame('ko' = kos, 'name' = ko_name, 'description' = ko_description, 'module_id' = module_id, 'module' = module_name, 'pathway_id' = map_id, 'pathway' = map_name, 'class' = map_class)
          res <- rbind(res, temp)
        }
      }
    }
  }else{
    next
  }
}

write.table(res, 'kos_API.txt', quote = F, sep = '\t', row.names = F)

File: KEGG/keggAPI.py
#!/usr/bin/env python
# encoding: utf-8
"""
@author: zhuhz
@file: keggAPI.py
@time: 2020/6/16 13:58
"""

from Bio.KEGG import REST

pathways = REST.kegg_list('pathway').read()
# print(type(pathways))
res = 'ko' + '\t' + 'ko name' + '\t' + 'ko des' + '\t' + 'module' + '\t' + 'module name' + '\t' + 'map' + '\t' + 'map name' + '\t' + 'map class' + '\n'

for pathway in pathways.rstrip().split('\n'):
	map_entry, map_description = pathway.split('\t')
	# print(map_entry)
	map = REST.kegg_get(map_entry).read()
	# print(type(map))
	current_section = None
	for line in map.rstrip().split('\n'):
		section = line[:12].strip()
		if not section == '':
			current_section = section
		if current_section == 'NAME':
			map_name = line[12:]
		elif current_section == 'CLASS':
			map_class = line[12:]
		# elif current_section == 'PATHWAY_MAP':
		# 	map_des = line[22:]
		elif current_section == 'MODULE':
			modules = line[12:]
			for module in modules.rstrip().split('\n'):
				module_id, module_name = module.split('  ')
				module_entry = 'module:' + module_id
				module_info = REST.kegg_get(module_entry).read()
				module_section = None
				for module_line in module_info.rstrip().split('\n'):
					module_section = module_line[:12].strip()
					if not module_section == '':
						c_module_section = module_section
					if c_module_section == 'ORTHOLOGY':
						orthologies = module_line[12:]
						ko_list = []
						for orth in orthologies.rstrip().split('\n'):
							try:
								kos, ko_des = orth.split('  ')
							except ValueError:
								kos, ko_des = orth.split(' ', 1)
							if len(kos) > 6:
								if kos[6] == '+':
									if len(kos) > 13 and kos[13] == ',':
										ko_list.append(kos.split('+')[0])
										ko_list.extend(kos.split('+')[1].split(','))
									elif len(kos) > 35 and len(kos.split('+')[4]) > 6 and kos.split('+')[4][6] == '-':
										ko_list.extend(kos.split('+')[:4])
										ko_list.extend(kos.split('+')[4].split('-'))
										ko_list.append(kos.split('+')[5])
									else:
										ko_list.extend(kos.split('+'))
								else:
									ko_list.extend(kos.split(','))
							else:
								ko_list.append(kos)
							for ko in ko_list:
								ko_entry = 'orthology:' + ko
								ko_info = REST.kegg_get(ko_entry).read()
								ko_section = None
								for ko_line in ko_info.rstrip().split('\n'):
									ko_section = ko_line[:12].strip()
									if not ko_section == '':
										c_ko_section = ko_section
									if c_ko_section == 'NAME':
										ko_name = ko_line[12:]
									elif current_section == 'DEFINITION':
										ko_des = ko_line[12:]
									else:
										continue
									temp = ko + '\t' + ko_name + '\t' + ko_des + '\t' + module_id + '\t' + module_name + '\t' + map_entry + '\t' + map_name + '\t' + map_class + '\n'
									res += temp
		else:
			continue

with open('kos_pyAPI.txt', 'w') as f:
	f.write(res)
